tools/src/main/java/org/apache/kafka/tools/VerifiableConsumer.java

```diff
@@ -0,0 +1,609 @@
+ *     See {@link org.apache.kafka.tools.VerifiableConsumer.PartitionsRevoked}</li>
+ * <li>partitions_assigned: outputs the partitions assigned through {@link ConsumerRebalanceListener#onPartitionsAssigned(Collection)}
+ *     See {@link org.apache.kafka.tools.VerifiableConsumer.PartitionsAssigned}.</li>
+ * <li>records_consumed: contains a summary of records consumed in a single call to {@link KafkaConsumer#poll(long)}.
+ *     See {@link org.apache.kafka.tools.VerifiableConsumer.RecordsConsumed}.</li>
+ * <li>record_data: contains the key, value, and offset of an individual consumed record (only included if verbose
+ *     output is enabled). See {@link org.apache.kafka.tools.VerifiableConsumer.RecordData}.</li>
+ * <li>offsets_committed: The result of every offset commit (only included if auto-commit is not enabled).
+ *     See {@link org.apache.kafka.tools.VerifiableConsumer.OffsetsCommitted}</li>
+ * </ul>
+ */
+public class VerifiableConsumer implements Closeable, OffsetCommitCallback, ConsumerRebalanceListener {
+
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final PrintStream out;
+    private final KafkaConsumer<String, String> consumer;
+    private final String topic;
+    private final boolean useAutoCommit;
+    private final boolean useAsyncCommit;
+    private final boolean verbose;
+    private final int maxMessages;
+    private int consumedMessages = 0;
+
+    private CountDownLatch shutdownLatch = new CountDownLatch(1);
+
+    public VerifiableConsumer(KafkaConsumer<String, String> consumer,
+                              PrintStream out,
+                              String topic,
+                              int maxMessages,
+                              boolean useAutoCommit,
+                              boolean useAsyncCommit,
+                              boolean verbose) {
+        this.consumer = consumer;
+        this.out = out;
+        this.topic = topic;
+        this.maxMessages = maxMessages;
+        this.useAutoCommit = useAutoCommit;
+        this.useAsyncCommit = useAsyncCommit;
+        this.verbose = verbose;
+        addKafkaSerializerModule();
+    }
+
+    private void addKafkaSerializerModule() {
+        SimpleModule kafka = new SimpleModule();
+        kafka.addSerializer(TopicPartition.class, new JsonSerializer<TopicPartition>() {
+            @Override
+            public void serialize(TopicPartition tp, JsonGenerator gen, SerializerProvider serializers) throws IOException {
+                gen.writeStartObject();
+                gen.writeObjectField("topic", tp.topic());
+                gen.writeObjectField("partition", tp.partition());
+                gen.writeEndObject();
+            }
+        });
+        mapper.registerModule(kafka);
+    }
+
+    private boolean hasMessageLimit() {
+        return maxMessages >= 0;
+    }
+
+    private boolean isFinished() {
+        return hasMessageLimit() && consumedMessages >= maxMessages;
```

Comment:
I wonder if you might want a timeout mechanism?

For example, if for some reason the number of messages in the topic is less than `maxMessages`, it seems your message fetch loop could spin forever.